### PR TITLE
chore (docs): add migration guide from rsc to ui

### PR DESCRIPTION
**Pull request by jeremyphilemon**

### CODE REVIEW

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 1]
```diff
@@ -0,0 +1,642 @@
...
+On the other hand, the AI SDK UI uses the `messages` field of `useChat` to store messages. To load messages when `useChat` is mounted, you should use `initialMessages`.
+
+Since messages are typically loaded from the database, we can use a server actions inside a Page component to fetch an older chat from the database during static generation and pass the messages as props to the `<Chat/>` component.
+
+#### After: Load chat from database during static generation of page
+
+```tsx filename="@/app/chat/[id]/page.tsx"
+import { Chat } from '@/app/components/chat';
+import { getChatById } from '@/utils/queries';
+import { convertToUIMessages } from '@/utils/functions';
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:471]: we probably need to define and then link to this somewhere

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 2]
```diff
@@ -0,0 +1,642 @@
...
+    </div>
+  );
+}
+```
+
+This setup gets your chat-based application up and running!
+
+### Generative User Interfaces
+
+The `streamUI` function uses `tools` as a wat to execute functions based on user input and renders React components based on the function output to go beyond text in the chat interface.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:165]:
```suggestion
The `streamUI` function uses `tools` as a way to execute functions based on user input and renders React components based on the function output to go beyond text in the chat interface.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 3]
```diff
@@ -0,0 +1,642 @@
...
+
+# Migration Guide (RSC to UI)
+
+This guide helps you migrate from AI SDK RSC to AI SDK UI.
+
+## Background
+
+The AI SDK has two packages that help you build the frontend for your applications – [AI SDK UI](/docs/ai-sdk-ui) and [AI SDK RSC](/docs/ai-sdk-rsc).
+
+We introduced support for using [React Server Components](https://react.dev/reference/rsc/server-components) (RSC) within the AI SDK to simplify building generative user interfaces for frameworks that supported RSC.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:14]:
```suggestion
We introduced support for using [React Server Components](https://react.dev/reference/rsc/server-components) (RSC) within the AI SDK to simplify building generative user interfaces for frameworks that support RSC.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 4]
```diff
@@ -0,0 +1,642 @@
...
+
+However, since we're pushing the boundaries of this technology, AI SDK RSC currently faces significant limitations that make it unsuitable for stable production use.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:16]:
```suggestion
However, given we're pushing the boundaries of this technology, AI SDK RSC currently faces significant limitations that make it unsuitable for stable production use.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 5]
```diff
@@ -0,0 +1,642 @@
...
+
+- It is not possible to abort a stream using server actions. This will be improved in future releases of React and Next.js [(1122)](https://github.com/vercel/ai/issues/1122).
+- When using `createStreamableUI` and `streamUI`, components remount on `.done()`, causing them to flicker [(2939)](https://github.com/vercel/ai/issues/2939).
+- Large amount of suspense can lead to crashes [(2843)](https://github.com/vercel/ai/issues/2843).
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:20]:
```suggestion
- Many suspense boundaries can lead to crashes [(2843)](https://github.com/vercel/ai/issues/2843).
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 6]
```diff
@@ -0,0 +1,642 @@
...
+- Using `createStreamableUI` can lead to quadratic data transfer. You can avoid this using createStreamableValue instead, and rendering the component client-side.
+- Closed RSC streams cause update issues [(3007)](https://github.com/vercel/ai/issues/3007).
+
+Due to these limitations, AI SDK RSC is marked as experimental, and we do not recommend using it for stable production environments.
+
+As a result, we strongly recommend migrating to AI SDK UI, which has undergone extensive development to provide a more stable and production grade experience.
+
+By building [v0](https://v0.dev) over the past year, we spent a lot of time figuring out ways to deliver the best chat experience for the web. The AI SDK UI packs some of these best practices and commonly used patterns like middleware, multi-step tool calls, attachments, telemetry, provider registry, and many more into a neat abstraction that you can use to reliably integrate AI into your applications.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:28]:
```suggestion
In building [v0](https://v0.dev), we have invested considerable time exploring how to create the best chat experience on the web. AI SDK UI ships with many of these best practices and commonly used patterns like [language model middleware](/docs/ai-sdk-core/middleware), [multi-step tool calls](/docs/ai-sdk-core/tools-and-tool-calling#multi-step-calls), [attachments](/docs/ai-sdk-ui/chatbot#attachments-experimental), [telemetry](/docs/ai-sdk-core/telemetry), [provider registry](/docs/ai-sdk-core/provider-management#provider-registry), and many more. These features have been considerately designed into a neat abstraction that you can use to reliably integrate AI into your applications.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 7]
```diff
@@ -0,0 +1,642 @@
...
+    },
+  });
+
+  return stream;
+}
+```
+
+#### Before: Call server action and update UI state
+
+The chat interface calls the server action, whose response is then stored in the `useUIState` hook.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:68]:
```suggestion
The chat interface calls the server action. The response is then saved using the `useUIState` hook.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 8]
```diff
@@ -0,0 +1,642 @@
...
+      >
+        <input type="text" />
+        <button type="submit">Submit</button>
+      </form>
+    </div>
+  );
+}
+```
+
+The `streamUI` function combines generating text and rendering the user interface. To migrate to AI SDK UI, you need to separate these concerns – streaming generations with `streamText` and rendering the UI with `useChat`.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:99]: This is really well put. Maybe worth bolding?

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 9]
```diff
@@ -0,0 +1,642 @@
...
+
+#### After: Replace server action with route handler
+
+The `streamText` function executes as part of a route handler and streams the response to the client. The `useChat` hook on the client decodes this stream and renders the chat interface.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:103]:
```suggestion
The `streamText` function executes as part of a route handler and streams the response to the client. The `useChat` hook on the client decodes this stream and renders the response within a chat interface.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 10]
```diff
@@ -0,0 +1,642 @@
...
+          }}
+        />
+        <button type="submit">Send</button>
+      </form>
+    </div>
+  );
+}
+```
+
+This setup gets your chat-based application up and running!
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:161]: You could cut this

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 11]
```diff
@@ -0,0 +1,642 @@
...
+        <button type="submit">Send</button>
+      </form>
+    </div>
+  );
+}
+```
+
+### Handling Client Interactions
+
+With AI SDK RSC, components streamed to the client become interactive by using event handlers to call server actions. This is possible as long as the component is a descendant of the `<AI/>` context provider.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:305]:
```suggestion
With AI SDK RSC, components streamed to the client can trigger subsequent generations by calling the relevant server action using the `useActions` hooks. This is possible as long as the component is a descendant of the `<AI/>` context provider.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 12]
```diff
@@ -0,0 +1,642 @@
...
+        </div>
+      ))}
+    </div>
+  );
+}
+```
+
+#### After: Use another chat hook with same ID from the component
+
+After switching to AI SDK UI, these messages are by initializing the `useChat` hook in the component with the same `id` as the parent component.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:341]: "these messages are by"?

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 13]
```diff
@@ -0,0 +1,642 @@
...
+        </div>
+      ))}
+    </div>
+  );
+}
+```
+
+### Saving Chats
+
+Before implementing `streamUI` as a server action, you should create an `<AI/>` provider and wrap your application at the root layout to sync between AI and UI states. During its initialization, you typically use the `onSetAIState` callback function to track updates of a mutable AI state and save it to the database when `done(...)` is called.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:377]:
```suggestion
Before implementing `streamUI` as a server action, you should create an `<AI/>` provider and wrap your application at the root layout to allow you to sync your AI and UI states. During initialization, you typically use the `onSetAIState` callback function to track updates to the AI State and save it to the database when `done(...)` is called.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 14]
```diff
@@ -0,0 +1,642 @@
...
+    if (done) {
+      await saveChat(state);
+    }
+  },
+});
+```
+
+#### After: Save chats using callback function of `streamText`
+
+When you switch over to AI SDK UI, you will save chats using the `onFinish` callback function of `streamText` present in the route handler.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:403]:
```suggestion
With AI SDK UI, you can save chats using the `onFinish` callback function of `streamText` in your route handler.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 15]
```diff
@@ -0,0 +1,642 @@
...
+
+  const result = await streamText({
+    model: openai('gpt-4o'),
+    system: 'you are a friendly assistant!',
+    messages: coreMessages,
+    onFinish: async ({ responseMessages }) => {
+      try {
+        await saveChat({
+          id,
+          messages: [...coreMessages, ...responseMessages],
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:423]: what are `responseMessages?`

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 16]
```diff
@@ -0,0 +1,642 @@
...
+
+    const chat = await loadChatFromDB();
+    const uiState = convertToUIState(chat);
+
+    return uiState;
+  },
+});
+```
+
+On the other hand, the AI SDK UI uses the `messages` field of `useChat` to store messages. To load messages when `useChat` is mounted, you should use `initialMessages`.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:462]:
```suggestion
AI SDK UI uses the `messages` field of `useChat` to store messages. To load messages when `useChat` is mounted, you should use `initialMessages`.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 17]
```diff
@@ -0,0 +1,642 @@
...
+
+Since messages are typically loaded from the database, we can use a server actions inside a Page component to fetch an older chat from the database during static generation and pass the messages as props to the `<Chat/>` component.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:464]:
```suggestion
As messages are typically loaded from the database, we can use a server actions inside a Page component to fetch an older chat from the database during static generation and pass the messages as props to the `<Chat/>` component.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 18]
```diff
@@ -0,0 +1,642 @@
...
+        </div>
+      ))}
+    </div>
+  );
+}
+```
+
+## Streaming Object Generation
+
+The `createStreamableValue` function streams any serializable data from the server to the client. As a result, this function streams object generations when used along with `streamObject`.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:521]:
```suggestion
The `createStreamableValue` function allows you to stream any serializable data from the server to the client. As a result, this function allows you to streams object generations from the server to the client when used along with `streamObject`.
```

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx [hunk 19]
```diff
@@ -0,0 +1,642 @@
...
+        }}
+      >
+        Generate
+      </button>
+    </div>
+  );
+}
+```
+
+To migrate to AI SDK UI, you should use the `useObject` hook and implement `streamObject` as part of a route handler.
```
Comment (nicoalbanese) [inline content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:590]:
```suggestion
To migrate to AI SDK UI, you should use the `useObject` hook and implement `streamObject` within your route handler.
```

**Review by nicoalbanese — APPROVED** [review]

**Review by lgrammel — APPROVED** [review]

**Comment by lgrammel** [comment]
Instead of putting this under troubleshooting, can we add it to the bottom of the AI SDK RSC section and also update the messages to point to the migration guide?

File: content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx (hunk 15 above)
Comment (jeremyphilemon) [reply content/docs/08-troubleshooting/01-migration-guide/04-migration-guide-rsc-ui.mdx:423]: https://sdk.vercel.ai/docs/ai-sdk-core/tools-and-tool-calling#response-messages

**Pull request merged**